bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/SWTFontProvider.java

```diff
@@ -26,20 +27,45 @@
  * take the provided values for the zoom into consideration and return scaled variant of a font if necessary.
  */
 public class SWTFontProvider {
-	private static Map<Device, SWTFontRegistry> fontRegistries = new HashMap<>();
+	private static Map<Device, SWTFontRegistry> fontRegistries = new ConcurrentHashMap<>();
 
 	private static SWTFontRegistry getFontRegistry(Device device) {
 		return fontRegistries.computeIfAbsent(device, SWTFontProvider::newFontRegistry);
 	}
 
+	/**
+	 * Returns the system font for the given device at the specified zoom.
+	 *
+	 * <b>Note:</b> This operation is not thread-safe. It must thus always be called
+	 * from the same thread for the same device, such as the display's UI thread.
```

Comment:
It seems odd to say "it's not thread-safe" while making it thread-safe (implementation detail). How about something more relaxed, like not _guaranteeing_ thread-safety?:

```suggestion
	 * <b>Note:</b> This operation is not guaranteed to be thread-safe. It should thus always be called
	 * from the same thread for the same device, such as the display's UI thread.
```

Reply:
I don't get it. Those methods are not thread-safe and will not be made thread-safe by this change, so what is wrong about saying that they are not thread-safe and instead relaxing it to "not guaranteed to be thread-safe"?

Reply:
The methods might not be synchronized but they will work just fine for multi-threaded apps calling them with the same device from different threads (because of the usage of `ConcurrentHashMap`).

Reply:
But that's not the meaning "thread-safe".

Reply:
I go by [this one](https://en.wikipedia.org/wiki/Thread_safety)

> In multi-threaded computer programming, a function is thread-safe when it can be invoked or accessed concurrently by multiple threads without causing unexpected behavior, race conditions, or data corruption.

... which doesn't mention at which level the thread-safeness is guaranteed. 

Which definition do you use?

Reply:
Exactly that definition is not fulfilled. What is a "thread-safeness level"?

Reply:
@fedejeanne the cache lookup is save, but the method being called from the cached object might not! Thats how far I understand the issue here...

Reply:
Exactly. And we have already agreed that we can accept this as this class is internal and the methods are (currently) only called from one thread per device (like the display's UI thread). That's why even though the method is not thread-safe, currently no concurrency issues may occur. And that's also why we agreed to add this documentation, as further usages of these methods could of course introduce concurrency issues like race conditions if performed in other threads.

Reply:
> ... the cache lookup is save, but the method being called from the cached object might not!
@laeubi you're right, `getSystemFont(zoom)` is not thread-safe. Thank you!